### PR TITLE
ABC v6 export

### PR DESCRIPTION
**Pull request by Five-Damned-Dollarz**
Could use some optimization on the export still, but should be 100% functional right now.
Your armature needs to be rotated [90, 0, 0], and scaled [-1, 1, -1] for the export to look correct.
You need to "calculate inside" in Blender to wind the triangles correctly (at least that's what has worked reliably for me).

**Known Issues**
- Using scaling will export incorrect bounding boxes.

**Complete**
- Geometry, normals, UV map
- Skeletal animation
- Vertex animation
- Bounding boxes

**In Progress**

**Not Started**
- Lithtech 1.5 TransformInfo

### CODE REVIEW

**Comment by Five-Damned-Dollarz** [comment]
This should be complete now, ~~unless I've somehow missed a huge bug~~ except for a small bug that affects _some_ vertex animated nodes that puts them a little off. Seems to be consistent, but I don't have a method to replicate it in a fresh .blend yet.
There's at least one derpy workaround in there, and the export process could probably be optimized, but as far as I can see it works great. Importing then exporting an original game file looks virtually identical (except for bounding boxes, which I try to create as best I can from the geometry).
The one thing I'm concerned about is node bounding boxes, since I haven't found any debug commands in Blood 2 to display them on models, I can't be sure they're correct. We can make changes easily enough in the future if it becomes clear they need fixing.

Edit (21/08/17): just for reference for myself/anyone later, the node bounding boxes _do not matter!_ The engine doesn't use them, or expose them to the server/client code at all. They can be ignored entirely for LithTech 1 use.

File: src/abc.py [hunk 1]
```diff
@@ -71,6 +71,9 @@ class Face(object):
     def __init__(self):
         self.vertices = []
 
+        # ABCv6 specific
+        self.normal=Vector()
```
Comment (haekb) [inline src/abc.py]:
```suggestion
        self.normal = Vector()
```

File: src/abc.py [hunk 2]
```diff
@@ -138,9 +141,12 @@ def __init__(self):
         self.child_count = 0
 
         # Version 6 specific
+        self.bounds_min=Vector()
```
Comment (haekb) [inline src/abc.py]:
```suggestion
        self.bounds_min = Vector()
```

File: src/abc.py [hunk 3]
```diff
@@ -138,9 +141,12 @@ def __init__(self):
...
+        self.bounds_max=Vector()
```
Comment (haekb) [inline src/abc.py]:
```suggestion
        self.bounds_max = Vector()
```

File: src/abc.py [hunk 4]
```diff
@@ -197,17 +203,19 @@ def __init__(self):
         self.node_keyframe_transforms = []
 
         # Version 6 specific
-
+        self.bounds_min=Vector()
```
Comment (haekb) [inline src/abc.py]:
```suggestion
        self.bounds_min = Vector()
```

File: src/abc.py [hunk 5]
```diff
@@ -197,17 +203,19 @@ def __init__(self):
...
+        self.bounds_max=Vector()
```
Comment (haekb) [inline src/abc.py]:
```suggestion
        self.bounds_max = Vector()
```

File: src/abc.py [hunk 6]
```diff
@@ -197,17 +203,19 @@ def __init__(self):
...
         # Note this should line up with md_vert_list, and go on for md_vert_count * keyframe_count
         # List of 3 chars (verts)
         self.vertex_deformations = []
+        self.vertex_deformation_bounds=dict()
```
Comment (haekb) [inline src/abc.py]:
```suggestion
        self.vertex_deformation_bounds = dict()
```

File: src/abc.py [hunk 7]
```diff
@@ -197,17 +203,19 @@ def __init__(self):
...
         # List of Vector (verts)
         # Scaled by the animation bounding box
         self.transformed_vertex_deformations = []
 
         # LTB specific
         self.compression_type = 0
-        self.is_vetex_animation = 0
+        self.is_vertex_animation = 0
```
Comment (haekb) [inline src/abc.py:218]: Thanks!

File: src/builder.py [hunk 8]
```diff
@@ -117,6 +119,8 @@ def from_armature(armature_object):
                     raise Exception('Mesh \'{}\' is not triangulated.'.format(
                         mesh.name))  # TODO: automatically triangulate the mesh, and have this be reversible
                 face = Face()
+                face.normal=polygon.normal
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                face.normal = polygon.normal
```

File: src/builder.py [hunk 9]
```diff
@@ -152,6 +156,15 @@ def from_armature(armature_object):
 
             node.bind_matrix = matrix
 
+            # ABC v6 specific
+            # FIXME: disgusting, can this be done better?
+            for piece in model.pieces:
+                for lod in piece.lods:
+                    for vertex in lod.vertices:
+                        if vertex.weights[0].node_index==bone_index:
+                            node.bounds_min=Vector((min(node.bounds_min.x, vertex.location.x), min(node.bounds_min.y, vertex.location.y), min(node.bounds_min.z, vertex.location.z)))
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                            node.bounds_min = Vector((min(node.bounds_min.x, vertex.location.x), min(node.bounds_min.y, vertex.location.y), min(node.bounds_min.z, vertex.location.z)))
```

File: src/builder.py [hunk 10]
```diff
@@ -152,6 +156,15 @@ def from_armature(armature_object):
...
+                            node.bounds_max=Vector((max(node.bounds_min.x, vertex.location.x), max(node.bounds_min.y, vertex.location.y), max(node.bounds_min.z, vertex.location.z)))
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                            node.bounds_max = Vector((max(node.bounds_min.x, vertex.location.x), max(node.bounds_min.y, vertex.location.y), max(node.bounds_min.z, vertex.location.z)))
```

File: src/builder.py [hunk 11]
```diff
@@ -241,24 +261,42 @@ def from_armature(armature_object):
...
             for time in keyframe_timings[model.nodes[0].name]['rotation_quaternion']:
                 # Expand our time
-                scaled_time = time * (1.0/0.025)
+                scaled_time = time * (1.0/get_framerate())
+                bpy.context.scene.frame_set(time, subframe=time-floor(time))
 
                 keyframe = Animation.Keyframe()
                 keyframe.time = scaled_time
+
+                keyframe.bounds_min=Vector(mesh_object.bound_box[0]) # will using mesh_object here break if there's multiple mesh
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                keyframe.bounds_min = Vector(mesh_object.bound_box[0]) # will using mesh_object here break if there's multiple mesh
```

File: src/builder.py [hunk 12]
```diff
@@ -241,24 +261,42 @@ def from_armature(armature_object):
...
+                keyframe.bounds_max=Vector(mesh_object.bound_box[6]) # objects using the same armature as a parent? DON'T DO THAT
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                keyframe.bounds_max = Vector(mesh_object.bound_box[6]) # objects using the same armature as a parent? DON'T DO THAT
```

Comment (haekb) [reply src/builder.py]: What's the significance of [6]? Please add some context, thanks.

File: src/builder.py [hunk 13]
```diff
@@ -241,24 +261,42 @@ def from_armature(armature_object):
             # For now we can just use the first node!
             for time in keyframe_timings[model.nodes[0].name]['rotation_quaternion']:
                 # Expand our time
-                scaled_time = time * (1.0/0.025)
+                scaled_time = time * (1.0/get_framerate())
+                bpy.context.scene.frame_set(time, subframe=time-floor(time))
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                subframe_time = time - floor(time)
                bpy.context.scene.frame_set(time, subframe=subframe_time)
```
Makes it look a little cleaner

File: src/builder.py [hunk 14]
```diff
@@ -184,13 +197,20 @@ def from_armature(armature_object):
 
         ''' Animations '''
         for action in bpy.data.actions:
+            if match(r"^d_", action.name): # skip any actions prefixed with "d_"; they're vertex animation lanes, we don't want them in the ABCv6 output
```
Comment (haekb) [inline src/builder.py]:
```suggestion
            # skip any actions prefixed with "d_"; they're vertex animation lanes, we don't want them in the ABCv6 output
            if match(r"^d_", action.name): 
```

File: src/builder.py [hunk 15]
```diff
@@ -184,13 +197,20 @@ def from_armature(armature_object):
...
             animation = Animation()
             animation.name = action.name
-            animation.extents = Vector((10, 10, 10))
 
             armature_object.animation_data.action = action
 
+            # TODO: make this optional for when we don't use vertex animation on anything
+            #if mesh.shape_keys:
+            #    md_action=[md_action for md_action in bpy.data.actions if md_action.name=="d_" + action.name]
+            #    mesh.shape_keys.animation_data.action=md_action[0] if md_action else None
```
Comment (haekb) [inline src/builder.py]: Do you still need this code?

File: src/builder.py [hunk 16]
```diff
@@ -241,24 +261,42 @@ def from_armature(armature_object):
...
 
                 keyframe = Animation.Keyframe()
                 keyframe.time = scaled_time
+
+                keyframe.bounds_min=Vector(mesh_object.bound_box[0]) # will using mesh_object here break if there's multiple mesh
+                keyframe.bounds_max=Vector(mesh_object.bound_box[6]) # objects using the same armature as a parent? DON'T DO THAT
+
                 animation.keyframes.append(keyframe)
 
+            # this is horrible, but works
```
Comment (haekb) [inline src/builder.py]:
```suggestion
```

File: src/builder.py [hunk 17]
```diff
@@ -241,24 +261,42 @@ def from_armature(armature_object):
...
-                
+
                 # FIXME: In the future we may trim off timing with no rotation/location changes
                 # So we'd have to loop through each keyframe timing, but for now this should work!
                 for time in keyframe_timing['rotation_quaternion']:
                     # Expand our time
-                    scaled_time = time * (1.0/0.025)
-                    bpy.context.scene.frame_set(time)
+                    scaled_time = time * (1.0/get_framerate())
+                    bpy.context.scene.frame_set(time, subframe=time-floor(time))
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                    subframe_time = time - floor(time)
                    bpy.context.scene.frame_set(time, subframe=subframe_time)
```

File: src/builder.py [hunk 18]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+
+        ''' Vertex Animations '''
+
+        # Then most trivially, you find min and max of each dimension, set scales to (maxes-mins), subtract mins from all points, then divide by scales.
+        # And the transform is set by doing the same subtract and divide to the origin
+        # Later on, to reduce artifacts, instead of just doing that and then scaling back up to 255 blindly, you could iterate over possible values UP to 255, and check sum of error^2 for each one, and choose the one with lowest total error
+        # Idea being that if you had like three evenly spaced things, then 240 may give you perfect accuracy, while 255 will not
+        # When you're compressing, you can choose to compress to less than 255, and just use a larger scale and transform to compensate
+
+        vertex_tolerance=1e-5
```
Comment (haekb) [inline src/builder.py]:
```suggestion
        vertex_tolerance = 1e-5
```

File: src/builder.py [hunk 19]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+
+        if mesh.shape_keys and len(mesh.shape_keys.key_blocks)>1:
+            for animation in model.animations:
+                print("Processing vertex animation", animation.name)
+
+                animation.vertex_deformations=dict()
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                animation.vertex_deformations = dict()
```

File: src/builder.py [hunk 20]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+
+                shape_keys=[shape_key for shape_key in mesh.shape_keys.key_blocks if shape_key.name.startswith(animation.name)]
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                shape_keys = [shape_key for shape_key in mesh.shape_keys.key_blocks if shape_key.name.startswith(animation.name)]
```

File: src/builder.py [hunk 21]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+
+                for node_index, node in enumerate(model.nodes):
+                    dirty_node=False
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                    dirty_node = False
```

File: src/builder.py [hunk 22]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+                    animation.vertex_deformations[node]=[]
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                    animation.vertex_deformations[node] = []
```

File: src/builder.py [hunk 23]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+
+                    # get all vertices for this node
+                    node_vertices=[vertex_index for vertex_index, vertex in enumerate(model.pieces[0].lods[0].vertices) if vertex.weights[0].node_index==node_index]
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                    node_vertices = [vertex_index for vertex_index, vertex in enumerate(model.pieces[0].lods[0].vertices) if vertex.weights[0].node_index == node_index]
```

File: src/builder.py [hunk 24]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+
+                    node.bounds_min=Vector((float("inf"), float("inf"), float("inf")))
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                    node.bounds_min = Vector((float("inf"), float("inf"), float("inf")))
```

File: src/builder.py [hunk 25]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+                    node.bounds_max=Vector((float("-inf"), float("-inf"), float("-inf")))
```
Comment (haekb) [inline src/builder.py]:
```suggestion
                    node.bounds_max = Vector((float("-inf"), float("-inf"), float("-inf")))
```

File: src/builder.py [hunk 26]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+                    if len(node_vertices)==0:
+                        node.bounds_min=Vector()
+                        node.bounds_max=Vector()
+
+                    # get the bounds of the node
+                    for keyframe_index, keyframe in enumerate(animation.keyframes):
+                        for vertex_index in node_vertices:
+                            temp_vert=shape_keys[keyframe_index].data[vertex_index]
+
+                            #if not temp_vert.co==mesh.shape_keys.key_blocks[0].data[vertex_index].co:
```
Comment (haekb) [inline src/builder.py]: You can probably remove this now

File: src/builder.py [hunk 27]
```diff
@@ -277,12 +315,96 @@ def from_armature(armature_object):
...
+                            break
+
+            if node.md_vert_count>0:
+                node.flags|=4
 
         ''' AnimBindings '''
         anim_binding = AnimBinding()
         anim_binding.name = 'base'
-        animation.extents = Vector((10, 10, 10))
+        anim_binding.extents = Vector((10, 10, 10))
```
Comment (haekb) [inline src/builder.py:404]: Thanks

**Review by haekb — CHANGES_REQUESTED** [review]
At some point github jammed up because I was adding too many suggestions. Please use spaces (and fix the ones I didn't suggest toooo!) 

Aside from that, I didn't notice any breakage from abc v12 -> lta export. So it's good to go.

**Comment by Five-Damned-Dollarz** [comment]
> At some point github jammed up because I was adding too many suggestions. Please use spaces (and fix the ones I didn't suggest toooo!)
> 
> Aside from that, I didn't notice any breakage from abc v12 -> lta export. So it's good to go.

I fixed as many as I could find, hopefully that's everything.

There is still one bug with the vertex animated node exporting with a slightly off rotation in rare animations, but it happens so infrequently in my experience that we could just make an issue for it, and hunt it later (with more information, hopefully, I have no idea where it comes from yet)

**Review by haekb — APPROVED** [review]

**Comment by haekb** [comment]
Cool beans, thanks. If you could just open a github issue about that, so we don't forget, that'll be greeeeat.

**Pull request merged**